external/fv3fit/fv3fit/pytorch/cyclegan/modules.py

```diff
@@ -144,6 +143,199 @@ def single_tile_convolution(
+        in_channels=in_channels,
+        out_channels=out_channels,
+        kernel_size=kernel_size,
+        output_padding=output_padding,
+        stride=stride,
+        stride_type=stride_type,
+        bias=bias,
+    )
+    if stride_type == "transpose":
+        # have to crop halo points from the output, as pytorch has no option to
+        # only output a subset of the domain for ConvTranspose2d
+        conv = nn.Sequential(
+            conv, Crop(n_halo=padding * stride + int(kernel_size - 1) // 2)
+        )
+    return nn.Sequential(append, conv)
+
+
+def cpu_only(method):
+    """
+    Decorator to mark a method as only being supported on the CPU.
+    """
+
+    @functools.wraps(method)
+    def wrapper(self, *args, **kwargs):
+        original_device = args[0].device
+        args = [arg.cpu() if isinstance(arg, torch.Tensor) else arg for arg in args]
+        kwargs = {
+            k: v.cpu() if isinstance(v, torch.Tensor) else v for k, v in kwargs.items()
+        }
+        return method(self, *args, **kwargs).to(original_device)
+
+    return wrapper
+
+
+class Crop(nn.Module):
+    def __init__(self, n_halo):
+        super(Crop, self).__init__()
+        self.n_halo = n_halo
+
+    @cpu_only
+    def forward(self, x):
+        return x[
+            ..., self.n_halo : -self.n_halo, self.n_halo : -self.n_halo
+        ].contiguous()
+
+
+class AppendHalos(nn.Module):
+
+    """
+    Module which appends horizontal halos to the input tensor.
+
+    Args:
+        n_halo: size of the halo to append
+    """
+
+    def __init__(self, n_halo: int):
+        super(AppendHalos, self).__init__()
+        self.n_halo = n_halo
+
+    def extra_repr(self) -> str:
+        return super().extra_repr() + f"n_halo={self.n_halo}"
+
+    @cpu_only
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            inputs: tensor of shape [batch, tile, channel, x, y]
+        """
+        corner = torch.zeros_like(inputs[:, 0, :, : self.n_halo, : self.n_halo])
+        if self.n_halo > 0:
+            with_halo = []
+            for _ in range(6):
+                tile = []
+                for _ in range(3):
+                    column = []
+                    for _ in range(3):
+                        column.append([None, None, None])  # row
+                    tile.append(column)
+                with_halo.append(tile)
+
+            for i_tile in range(6):
+                with_halo[i_tile][1][1] = inputs[:, i_tile, :, :, :]
+                with_halo[i_tile][0][0] = corner
+                with_halo[i_tile][0][2] = corner
+                with_halo[i_tile][2][0] = corner
+                with_halo[i_tile][2][2] = corner
+                # we must make data contiguous after rotating 90 degrees because
+                # the MPS backend doesn't properly manage strides when concatenating
+                # arrays
+                if i_tile % 2 == 0:  # even tile
+                    # south edge
```

Comment:
Can you clarify whether these comments use north/south/east/west in the true sense or relative to each tile's local x/y conventions?